### PR TITLE
potato udp hype

### PR DESCRIPTION
**Pull request by pdlla**
just a start on udp stuff 
also removed a bunch of old stuff

### CODE REVIEW

File: src/potato_records.hrl [hunk 1]
```diff
@@ -6,6 +6,13 @@
 -record(pending_tx, {player_map, counter}).
 
 
+-record(validator,
+  {gameid :: integer(),
+  pubkey :: public_key:public_key(),
+  address :: {string(), integer()}}).
```
Comment (antonbobkov) [inline src/potato_records.hrl]: lol i didn't know you can have types in records

File: src/p2p/potato_udp.erl [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+%% state maps {game_id} to game instance PID
+-spec init(integer()) -> {ok, {gen_udp:socket(), map()}}.
+init(Port) ->
+  io:format("gen_udp open on port: ~p~n", [Port]),
+  {ok, Socket} = gen_udp:open(Port, [binary, {active,true}]),
+  %% add to group?
+  {ok, {Socket, maps:new()}}.
+
+%% Never used
```
Comment (antonbobkov) [inline src/p2p/potato_udp.erl]: :sob:

File: src/p2p/potato_udp.erl [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+handle_call(_E, _From, S) ->
+  {noreply, S}.
+
+%% add a {game_id} => Pid to map
+handle_cast({add_game,Key,Pid}, S) ->
+  {Socket, Map} = S,
+  Map2 = maps:put(Key, Pid, Map),
+  {noreply, {Socket, Map2}};
+
```
Comment (antonbobkov) [inline src/p2p/potato_udp.erl]: you can also do `{add_game,Key,Pid}, {Socket, Map})` instead of `S`

File: src/p2p/potato_udp.erl [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      {noreply, S}
+  end;
+
+handle_info(E, S) ->
+  io:format("unexpected: ~p~n", [E]),
+  {noreply, S}.
+
+
+code_change(_OldVsn, State, _Extra) ->
+  {ok, State}.
```
Comment (antonbobkov) [inline src/p2p/potato_udp.erl]: :scream:

File: test/potato_udp_test.erl [hunk 5]
```diff
@@ -0,0 +1,12 @@
+-module(potato_udp_test).
+
+-include_lib("eunit/include/eunit.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+potato_udp_test() ->
+  Port = 3142,
+  {ok, Pid} = gen_server:start(potato_udp, Port, []),
+  %% ?debugFmt("started potato_udp gen_server ~p~n",[Pid]),
+  gen_server:cast(Pid, {send, "localhost", Port, "hi"}),
```
Comment (antonbobkov) [inline test/potato_udp_test.erl]: you forgot to pack before unpacking??

your unpack throws badarg exception see
https://travis-ci.org/antonbobkov/potato_game/builds/442478411?utm_source=github_status&utm_medium=notification

```
got: <<"hi">>
terminate reason: {badarg,
                      [{erlang,binary_to_term,[<<"hi">>],[]},
                       {messages,unpack,1,
```

Comment (antonbobkov) [reply test/potato_udp_test.erl]: should be `messages:pack("hi")` maybe?

or do it inside `handle_cast`

File: test/potato_udp_test.erl [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  ok = gen_server:stop(Pid).
+  %% TODO better tests
```
Comment (antonbobkov) [inline test/potato_udp_test.erl]: pff

File: src/p2p/potato_udp.erl [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+%% and then this function will handle formatting the message
+handle_cast({send, Address, Port, Msg}, S) ->
+  {Socket, _} = S,
+  gen_udp:send(Socket, Address, Port, Msg),
+  {noreply, S}.
+
+%% TODO log errors or whatever.
+handle_info({udp, _Socket, _IP, _InPortNo, Packet}, S) ->
+  io:format("got: ~p~n", [Packet]),
+  case messages:unpack(Packet) of
```
Comment (antonbobkov) [inline src/p2p/potato_udp.erl]: lol i also assumed you cannot have complex expressions inside of case

my code just suddenly got two times shorter

File: src/p2p/potato_udp.erl (hunk 3 above)
Comment (pdlla) [reply src/p2p/potato_udp.erl]: :O

File: src/p2p/potato_udp.erl (hunk 4 above)
Comment (pdlla) [reply src/p2p/potato_udp.erl]: 😱😱

File: test/potato_udp_test.erl (hunk 5 above)
Comment (pdlla) [reply test/potato_udp_test.erl]: 🙀

File: src/p2p/potato_udp.erl (hunk 7 above)
Comment (pdlla) [reply src/p2p/potato_udp.erl]: 👍 🥔

File: test/potato_udp_test.erl (hunk 5 above)
Comment (pdlla) [reply test/potato_udp_test.erl]: it's passing on my computer 😱

**Comment by pdlla** [comment]
in other news, I just discovered the drop down next to "merge pull request" where there is an option to rebase 🙀

**Comment by pdlla** [comment]
such linear history

File: src/p2p/potato_udp.erl (hunk 4 above)
Comment (antonbobkov) [reply src/p2p/potato_udp.erl]: :scream: :scream: :scream:

**Pull request merged**